webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx

```diff
@@ -36,508 +63,301 @@ type Props = PropsFromRedux & {
+        ignore_channel_mentions: ignoreChannelMentions,
+        channel_auto_follow_threads: channelMemberNotifyProps?.channel_auto_follow_threads || 'off',
+    };
+}
 
-        return currentUserNotifyProps.push;
-    }
-    return NotificationLevels.ALL;
+type SettingsType = {
+    desktop: ChannelNotifyProps['desktop'];
+    desktop_threads: ChannelNotifyProps['desktop_threads'];
+    mark_unread: ChannelNotifyProps['mark_unread'];
+    push: ChannelNotifyProps['push'];
+    push_threads: ChannelNotifyProps['push_threads'];
+    ignore_channel_mentions: ChannelNotifyProps['ignore_channel_mentions'];
+    channel_auto_follow_threads: ChannelNotifyProps['channel_auto_follow_threads'];
 };
 
-const getDefaultPushThreadsNotifyLevel = (currentUserNotifyProps: UserNotifyProps, isGM: boolean): Exclude<ChannelMemberNotifyProps['push_threads'], undefined> => {
-    if (currentUserNotifyProps?.push_threads) {
-        if (currentUserNotifyProps.push_threads === 'default') {
-            return NotificationLevels.ALL;
-        }
+export default function ChannelNotificationsModal(props: Props) {
+    const {formatMessage} = useIntl();
+    const [show, setShow] = useState(true);
+    const [serverError, setServerError] = useState('');
+    const [channelNotifyProps] = useState(props.channelMember && props.channelMember.notify_props);
+    const [mobileSettingsSameAsDesktop, setMobileSettingsSameAsDesktop] = useState<boolean>(getUseSameDesktopSetting(props.currentUser.notify_props, channelNotifyProps));
 
-        if (isGM && currentUserNotifyProps.push_threads === NotificationLevels.MENTION) {
-            return NotificationLevels.ALL;
-        }
+    const [settings, setSettings] = useState<SettingsType>(getStateFromNotifyProps(props.currentUser.notify_props, channelNotifyProps));
 
-        return currentUserNotifyProps.push_threads;
+    function handleHide() {
+        setShow(false);
     }
-    return NotificationLevels.ALL;
-};
-
-export default class ChannelNotificationsModal extends React.PureComponent<Props, State> {
-    constructor(props: Props) {
-        super(props);
 
-        const channelNotifyProps = props.channelMember?.notify_props;
-
-        this.state = {
-            show: true,
-            activeSection: NotificationSections.NONE,
-            serverError: null,
-            ...this.getStateFromNotifyProps(props.currentUser.notify_props, channelNotifyProps),
-        };
+    function handleExit() {
+        props.onExited();
     }
 
-    componentDidUpdate(prevProps: Props) {
-        const prevChannelNotifyProps = prevProps.channelMember && prevProps.channelMember.notify_props;
-        const channelNotifyProps = this.props.channelMember && this.props.channelMember.notify_props;
-
-        if (!deepEqual(channelNotifyProps, prevChannelNotifyProps)) {
-            this.resetStateFromNotifyProps(this.props.currentUser.notify_props, channelNotifyProps);
+    const resetToDefault = useCallback((settingName: string) => {
```

Comment:
This next suggestion is definitely out of scope for this PR, but I'm trying to think if there's some way we could wrap all this logic into a custom hook that:

1. Takes the default values for everything and possibly a function to handle saving changes
2. Keeps track of their current values and generates a type definition for them
3. Generates a more type safe version of `handleChange` that doesn't rely on `Record<string, string>`
4. Returns this `resetToDefault` callback

It would be nice if we could encapsulate all that logic in a way that we could more easily pass it into the new components you've added for when those are used elsewhere. That's definitely massive scope creep though 😅